utils/src/fiemap.c

```diff
@@ -0,0 +1,166 @@
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#include <linux/fiemap.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <argp.h>
+
+#include "sparse.h"
+#include "parse.h"
+#include "util.h"
+#include "format.h"
+#include "ioctl.h"
+#include "cmd.h"
+
+/*
+ * This is wholly modeled after e2fsprogs/filefrag.c from tso
+ */
+
+struct get_fiemap_args {
+	char *filename;
+	bool phys;
+	bool byte;
+};
+
+static int do_get_fiemap(struct get_fiemap_args *args)
+{
+	__u64 buf[2048];        /* __u64 for proper field alignment */
+	struct fiemap *fiemap = (struct fiemap *)buf;
+	struct fiemap_extent *fm_ext = &fiemap->fm_extents[0];
+	int count = (sizeof(buf) - sizeof(*fiemap)) /
+			sizeof(struct fiemap_extent);
+	int fd;
+	int ret;
+	int i;
+	int nr = 0; /* XXX we could put this in fm_start to make start/count an option */
+	int last = 0;
+	int written = 0;
+	u64 off;
+	u64 len;
+
+	memset(fiemap, 0, sizeof(struct fiemap));
+
+	fd = open(args->filename, O_RDONLY);
+	if (fd < 0) {
+		ret = -errno;
+		fprintf(stderr, "failed to open '%s': %s (%d)\n",
+			args->filename, strerror(errno), errno);
+		goto out;
+	}
+
+	do {
+		fiemap->fm_length = ~0ULL;
```

Comment:
This never sets fm_start so it'll loop forever if run on a file with enough extents to fill the buffer without encountering the _LAST flag.  I guess you'd want to track the next offset after the last read extent and put that in fm_start for the next call.

Reply:
Yes, seems that in rewriting the output section here I cut off the fm_start calculation. It's memset to 0 implicitly at the start for the first ioctl call, but then we need to recalculate it just before the end of the `do` loop. It's calculated as `fm_ext[i-1].fe_logical + fm_ext[i-1].fe_length` in the various copies of filefrag.c so that'll do.